src/test/spec/sessions/mod.rs

```diff
@@ -0,0 +1,287 @@
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+async fn run_sessions_tests() {
+    let _guard: RwLockWriteGuard<()> = LOCK.run_exclusively().await;
+    run_spec_test(&["sessions"], run_test).await;
+}
+
+async fn run_test(test_file: TestFile) {
+    let client = EventClient::new().await;
+
+    if let Some(requirements) = test_file.run_on {
+        let can_run_on = requirements.iter().any(|run_on| run_on.can_run_on(&client));
+        if !can_run_on {
+            println!("Client topology not compatible with test");
+            return;
+        }
+    }
+
+    for test in test_file.tests {
+        if test
+            .operations
+            .iter()
+            .any(|operation| SKIPPED_OPERATIONS.contains(&operation.name.as_str()))
+        {
+            continue;
+        }
+
+        if let Some(skip_reason) = test.skip_reason {
+            println!("skipping {}: {}", test.description, skip_reason);
+            continue;
+        }
+
+        match client
+            .database("admin")
+            .run_command(doc! { "killAllSessions": [] }, None)
+            .await
+        {
+            Ok(_) => {}
+            Err(err) => match err.kind.code_and_message() {
+                Some((11601, _)) => {}
+                _ => panic!("{}: killAllSessions failed", test.description),
+            },
+        }
+
+        let coll = client
+            .database(&test_file.database_name)
+            .collection(&test_file.collection_name);
+        let options = DropCollectionOptions::builder()
+            .write_concern(majority_write_concern())
+            .build();
+        coll.drop(options).await.unwrap();
+
+        let options = CreateCollectionOptions::builder()
+            .write_concern(majority_write_concern())
+            .build();
+        client
+            .database(&test_file.database_name)
+            .create_collection(&test_file.collection_name, options)
+            .await
+            .unwrap();
+
+        if let Some(data) = &test_file.data {
+            let options = InsertManyOptions::builder()
+                .write_concern(majority_write_concern())
+                .build();
+            coll.insert_many(data.clone(), options).await.unwrap();
+        }
+
+        let client = EventClient::with_additional_options(
+            test.client_options.clone(),
+            None,
+            test.use_multiple_mongoses,
+            None,
+            false,
+        )
+        .await;
+
+        if let Some(fail_point) = test.fail_point {
+            let doc = fail_point.command;
+            client
+                .database("admin")
+                .run_command(doc, None)
+                .await
+                .unwrap();
+        }
+
+        let options = match test.session_options {
+            Some(ref options) => options.get("session0").cloned(),
+            None => None,
+        };
+        let mut session0 = Some(client.start_session(options).await.unwrap());
```

Comment:
These sessions need to be wrapped in options because one of the possible test operations is `endSession`, which in our case requires calling `drop` (i.e. taking ownership of the session) -- to get around the borrow checker, the session is taken out of the option when it's necessary to `drop` and unwrapped everywhere else. The unwraps are safe since the tests never use a session after it's been ended.

Reply:
Sounds good! We had to do a similar thing with the connection pool spec tests.